gpii/node_modules/preferencesServer/src/preferencesService.js

```diff
@@ -73,12 +124,270 @@ var fluid = fluid || require("infusion");
+        gpiiKeyMissing: "GPII key \"%gpiiKey\" does not exist",
+        gpiiKeyExisted: "GPII key \"%gpiiKey\" already exists",
+        noUpdateOnSnapset: "Cannot update:  GPII key \"%gpiiKey\" is a snapset"
+    });
+
+    /**
+     * Returns a preferences safe using it's id. If the safe is not found an error object
+     * is returned.
+     *
+     * @param  {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param  {String} prefsSafeId Prefs Safe ID
+     * @return {Promise}            Returns a promise resolving with the preferences safe,
+     * or rejecting with an error payload containing an error `message`.
+     */
+    gpii.preferencesServer.preferencesService.getPrefsSafe = function (dataStore, prefsSafeId) {
+        var promiseTogo = fluid.promise();
+        dataStore.findById(prefsSafeId).then(
+            function (data) {
+                if ($.isEmptyObject(data) || data.type !== "prefsSafe") {
+                    promiseTogo.reject({
+                        isError: true,
+                        errorCode: gpii.preferencesServer.errors.missingPrefsSafe.errorCode,
+                        message: gpii.preferencesServer.errors.missingPrefsSafe.message
+                    });
+                }
+                else {
+                    promiseTogo.resolve(data);
+                }
+            },
+            function (err) {
+                fluid.log("Error looking up prefsSafe: ", err);
+                promiseTogo.reject({
+                    isError: true,
+                    errorCode: gpii.preferencesServer.errors.errorAccessingPrefsSafe.errorCode,
+                    message: gpii.preferencesServer.errors.errorAccessingPrefsSafe.message
+                });
+            }
+        );
+        return promiseTogo;
+    };
+
+    /**
+     * Adds a new preferences safe to CouchDB using the preferences safe format
+     * minus the `id`, `timestampCreated`, and `timestampUpdated` properties.
+     *
+     * @param {Object} that - Instance of `gpii.preferencesServer.preferencesService`.
+     * @param {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param {Object} prefsSafeData - Preferences safe to add to system.
+     * @return {Promise} Returns a promise resolving with the newly created
+     * preferences safe, or rejecting with an error payload containing an error `message`.
+     */
+    gpii.preferencesServer.preferencesService.addPrefsSafe = function (that, dataStore, prefsSafeData) {
+        var promiseTogo = fluid.promise();
+        if (prefsSafeData.type !== "prefsSafe") {
+            promiseTogo.reject({
+                isError: true,
+                errorCode: gpii.preferencesServer.errors.errorNotPrefsSafe.errorCode,
+                message: gpii.preferencesServer.errors.errorNotPrefsSafe.message
+            });
+        }
+        else {
+            dataStore.addPrefsSafe(prefsSafeData).then(
```

Comment:
Add a todo here that "prefsSafeData" should be validated when schema validation is implemented. The same comment applies to other functions that create or update documents.

Reply:
updated